stablehlo/dialect/TypeInference.h

```diff
@@ -0,0 +1,150 @@
+#include "mlir/Interfaces/InferTypeOpInterface.h"
+#include "mlir/Support/LogicalResult.h"
+
+namespace mlir {
+namespace hlo {
+
+// Check if the dimension size is dynamic.
+inline static bool isDynamicDimSize(int64_t val) {
+  return val == ShapedType::kDynamicSize;
+}
+
+// TODO(https://github.com/openxla/stablehlo/issues/270)
+// Remove the util functions below when all the shape functions are moved here.
+
+bool compatibleShapeAndElementType(Type type1, Type type2,
+                                   bool ignoreFpPrecision = false);
+
+FailureOr<SmallVector<int64_t>> convert1DAttribute(
+    Optional<DenseIntElementsAttr> optionalAttr, Optional<Location> loc,
+    StringRef attrName);
+
+FailureOr<SmallVector<std::pair<int64_t, int64_t>>> convertPaddingAttribute(
+    Optional<DenseIntElementsAttr> optionalAttr, Optional<Location> loc);
+
+// WindowDimension described how the kernel window moves across the base area
+// in a particular dimension.
+// Describes the windowing in an operation such as convolution.
+// The window is moved across a base area and for each position of the
+// window a computation is performed. The field below describes the
+// window and the movement of the window across a base area.
+struct WindowDimension {
+  int64_t size = 0;
+  int64_t stride = 1;
+  int64_t paddingLow = 0;
+  int64_t paddingHigh = 0;
+  int64_t windowDilation = 1;
+  int64_t baseDilation = 1;
+  bool windowReversal = false;
+};
+
+FailureOr<SmallVector<WindowDimension>>
+verifyWindowAttributesAndInferWindowDimensions(
+    ArrayRef<int64_t> windowDimensions, ArrayRef<int64_t> windowStrides,
+    ArrayRef<std::pair<int64_t, int64_t>> padding,
+    ArrayRef<int64_t> lhsDilation, ArrayRef<int64_t> rhsDilation,
+    Optional<Location> loc);
+
+SmallVector<int64_t> inferWindowOutputShape(
+    const ArrayRef<int64_t> baseShape, const ArrayRef<WindowDimension> window);
+
+unsigned potentiallyComplexBitwidth(Type type);
+
+//===----------------------------------------------------------------------===//
+// Shape functions for ops.
+//===----------------------------------------------------------------------===//
+
+LogicalResult inferBatchNormGradOp(
+    Value operand, uint64_t featureIndex,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes);
+
+LogicalResult inferBatchNormInferenceOp(
+    Value operand, SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes);
+
+LogicalResult inferBatchNormTrainingOp(
+    Value operand, uint64_t featureIndex,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes);
+
+LogicalResult inferCaseOp(
+    Optional<Location> location, RegionRange branches,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes);
+
+LogicalResult inferDotGeneralOp(
+    Optional<Location> location, Value lhs, Value rhs,
+    ArrayRef<int64_t> lhsBatchingDims, ArrayRef<int64_t> rhsBatchingDims,
+    ArrayRef<int64_t> lhsContractingDims, ArrayRef<int64_t> rhsContractingDims,
```

Comment:
Let's call these parameters `lhsBatchingDimensions`, `rhsBatchingDimensions`, `lhsContractingDimensions` and `rhsContractingDimensions` to match the ODS.

Reply:
Done.